pkg/pushsync/pushsync.go

```diff
@@ -3,3 +3,216 @@
 // license that can be found in the LICENSE file.
 
 package pushsync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/p2p/protobuf"
+	"github.com/ethersphere/bee/pkg/pushsync/pb"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/topology"
+)
+
+const (
+	protocolName    = "pushsync"
+	protocolVersion = "1.0.0"
+	streamName      = "pushsync"
+)
+
+type PushSync struct {
+	streamer      p2p.Streamer
+	storer        storage.Storer
+	peerSuggester topology.ClosestPeerer
+	quit          chan struct{}
+
+	logger  logging.Logger
+	metrics metrics
+}
+
+type Options struct {
+	Streamer      p2p.Streamer
+	Storer        storage.Storer
+	ClosestPeerer topology.ClosestPeerer
+
+	Logger logging.Logger
+}
+
+var retryInterval = 10 * time.Second // time interval between retries
+
+func New(o Options) *PushSync {
+	ps := &PushSync{
+		streamer:      o.Streamer,
+		storer:        o.Storer,
+		peerSuggester: o.ClosestPeerer,
+		logger:        o.Logger,
+		metrics:       newMetrics(),
+		quit:          make(chan struct{}),
+	}
+
+	ctx := context.Background()
```

Comment:
@janos WDYT about getting rid of `quit` in favor of keeping a `quitFunc`, that will just call the cancel function of this context?

Reply:
but then, should the context not be cancelled when the quit is called? since this context is being passed to other components. cancelling it would make sure any pending operations are cancelled

Reply:
In this case, there is no cancel function there is only a background context.

But this is a good point. This context is not doing any cancelation in this code, it is just shared without much purpose. I do not think that it is needed at all, and chunksWorker can create its own contexts where a function with a context should be called.